core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.options;
+
+import alluxio.grpc.CompleteFilePOptions;
+import alluxio.master.file.FileSystemMasterOptions;
+import alluxio.underfs.UfsStatus;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * Used to merge and wrap {@link CompleteFilePOptions}.
+ */
+public class CompleteFileContext extends OperationContext<CompleteFilePOptions.Builder> {
+
+  private long mOperationTimeMs;
+  private UfsStatus mUfsStatus;
+
+  /**
+   * Creates rename context with given option data.
+   *
+   * @param mergedOptionsBuilder options builder
+   */
+  private CompleteFileContext(CompleteFilePOptions.Builder mergedOptionsBuilder) {
+    super(mergedOptionsBuilder);
+    mOperationTimeMs = System.currentTimeMillis();
+    mUfsStatus = null;
+  }
+
+  /**
+   * Merges and embeds the given {@link CompleteFilePOptions} with the corresponding master options.
+   *
+   * @param optionsBuilder Builder for proto {@link CompleteFilePOptions} to merge with defaults
+   * @return the instance of {@link CompleteFileContext} with default values for master
+   */
+  public static CompleteFileContext defaults(CompleteFilePOptions.Builder optionsBuilder) {
```

Comment:
Can we call this `create`, `applyDefaults`, or `createWithDefaults`? We typically use verbs for methods that process an input.

Reply:
Agreed to naming.

Considering the overall comment you gave, on keeping the given defaults as is, how about keeping the proto options as builder but provide another method from base OperationContext to expose given defaults. So in the end we would have above three methods for creating the contexts and each context would have getOriginalOptions() method that returns whatever was provided by the caller before merging or modification. 

Using a builder makes setting proto fields very easy. It's not just to avoid extra .build() call at the end but to be able to set proto fields line-by-line as below, which is very usefull in places where there are many proto fields (e.g. InodeTree).

context = SomeContext.createWithDefaults();
context.getOptions().setQwe();
context.getOptions().setAsd();

compared to;

context = SomeContext.createWithDefaults();
context.setOptions(context.getOptions().toBuilder().setQwe().setAsd().build);

Reply:
At the start of the RPC we receive options from the client and merge them with the defaults to create a complete `*POptions` object. Most RPCs never modify this `*POptions` object. Ideally no RPCs would modify it. We can guarantee this by making it immutable, simplifying our lives greatly because we don't need to worry about all the options changing.

Modifying options during an RPC call is rare. I only see three places where we need to do it:

https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR807
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR890
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR921

All 3 of these are related to loadMetadataType/loadDescendantType, which are effectively deprecated now that we have ufs sync. We shouldn't base the convention off of these outliers. Immutability is valuable for preventing bugs, and we have an opportunity here to make dozens of fields (every option) immutable at a low cost code-wise

Reply:
@madanadit, this discussion is related to your question :
> should we set all options fields in one go?

Reply:
I put this on my task list for the consolidation work, as it'll cause lots of merge conflicts for pending PRs.